### ISSUE TITLE
Revision of the HDF5 Integration

### ISSUE THREAD

**Issue by rainman110** [dlr-gtlab/gtlab-core#173]
In GitLab by @mariusalexander on Dec 13, 2021, 14:07

# HDF5 specific Roadmap for DP4 

## Implement a robust interface to access externalized data
Currently the developer has to use great care when dealing with externalized objects:
 * The data must always be fetched before accessing it (using `GtExternalizedObject::fetchData()`)
 * The data must always be released after it is no longer needed (using `GtExternalizedObject::releaseData()`)
 * `fetchData()` and `releaseData()` must be called equally often

**Current pitfalls**
 * The data can be accessed eventhough it was no fetched yet:
```cpp
// datazone was not internalized here
dataZone->params(); // will return empty list!
```
 * Data can be overriden if the object was not fetched yet:
```cpp
// datazone was not internalized here
dataZone->setParams(QStringList()); // clearing params has no lasting effect
dataZone->fetchData(); // params are restored once fetched!
```
 * Object may not be externalized properly and thus create a "memory leak" (eg. `dataZone->releaseData()` was not called)
 * Object can be externalized multiple times and effect other processes:
```cpp
// datazone was already fetched before (eg. object is accessed by a Mdi view)
dataZone->fetchData(); // data was already fetched, internal ref count is now 2
dataZone->releaseData(); // data is still fetched, ref count is 1
...
dataZone->releaseData(); // data is no longer fetched, ref count is 0 -> Data in Mdi view may be corrupted!
```

**Idea:**
The developer should not be able to access the data if it is externalized. 
Addionally RAII should be used to manage the resource acquisition.

To archive this behaviour a helper object should be implemented that uses RAII, whichs fetchs the data in the constructor and frees/externalizes the data in the destructor. The data should only be accessible through the helper object:
```cpp
// old usage
dataZone->params(); // ERROR: no meber named params in GtDataZone0D
// new usage
GtDataZone0DData data = dataZone.fetchData(); // returns a helper object
data.params() // helper object provides access to the data
```

<details>
  <summary>Simple Class Hierarchy Proposal</summary>

___
```mermaid
classDiagram
    QObject <|-- GtObject
    GtObject <|-- GtExternalizedObject
    
    GtExternalizedObject <|-- GtAbstractDataZone
    GtAbstractDataZone <|-- GtDataZone
    GtAbstractDataZone <|-- GtDataZone0D

    
    GtExternalizedData <|-- GtAbstractDataZoneData
    GtAbstractDataZoneData <|-- GtDataZoneData
    GtAbstractDataZoneData <|-- GtDataZone0DData


    GtExternalizedData <.. GtExternalizedObject: Data can only be accesed\nusing data object instances

    class GtExternalizedObject{
      +int refCount
      +fetchData()
      +releaseData()
    }

    

    class GtAbstractDataZone{
      +QStringList params, units
    }

    class GtAbstractDataZoneData{
      +params()
      +units()
    }

    class GtDataZone0D{
      +QVector<double> values
    }

    class GtDataZone0DData{
      +values()
    }

    class GtDataZone{
      +GtTable ...
    }
    class GtDataZoneData{
      +values1D()
      +values2D()
      +...()
    }
```
___
</details>

## Rework the externalization behaviour

Currently a Enum is used to define the behaviour when externalizing:
* `Discard`: If the data was only fetched for read only purposes and may be cleared 
* `Externalize`: Update the data on disk if the data has changed (prevents repeated read/write operations)
* `ForceExternalize`: Force the externalization of the data

The Enum must always be provided when using the `GtExternalized::releaseData()` function. The object may therefore write multiple times to disk:
```cpp
// consider following scenario:
// the object is fetched and has a ref count of 3
GtExternalizedObject* obj(...);
obj->releaseData(GtExternalizedObject::Externalize); // refCount is 2, data was written to disk but was not cleared
...
obj->releaseData(GtExternalizedObject::Discard); // refCount is 1, no data was written to disk, data not cleared
...
obj->releaseData(GtExternalizedObject::Externalize); // refCount is 0, data was written to disk again, data cleared
```

**Idea:**
The object should only be externalized once the internal ref count is 0. This may prevent multiple write operations and the enum may be removed to simplify the API.

## Refactor the GtH5FileManager to be less HDF5 specific
The `GtH5FileManager` was implemented to allow the externalization to temporary files for objects, which were modified in a task/calculator. However the current implementation works only for `GtExternalizedh5Object`s. This behaviour should be generalized.  

## update the project.h5 file only once the project was saved
Currently the `project.h5` file contains all changes eventhough the project was not saved yet. A simple solution may be to use a temporary file, which will contain all changes and once the project is saved the changes will be applied to the `project.h5` file.

## Support Undo/Redo
Undo/Redo is not supported for externalized objects. Maybe the refactored `GtH5FileManager` can be used (e.g. use a temporary file for each new data version). 

## General improvements to the HDF5 Wrapper

* Use the Pimpl Idiom to abstract the implementation details (may clear dependencies to HDF5 libraries at compile time)
* Maybe use a shared/unique pointer for the `GtH5File` reference
* **Should the HDF5 Wrapper be made open source?**
